addons/account/tests/test_account_move_in_invoice.py

```diff
@@ -1564,13 +1564,69 @@ def test_in_invoice_write_1(self):
+        self.assertRecordValues(invoice_2, [{'duplicated_ref_ids': invoice_1.ids}])
+
+    def test_in_invoice_single_duplicate_supplier_reference_with_form(self):
+        """ Ensure duplicated ref are computed correctly with UI's NEW_ID"""
+        invoice_1 = self.invoice
+        invoice_1.ref = 'a unique supplier reference that will be copied'
+        move_form = Form(self.env['account.move'].with_context(default_move_type='in_invoice'))
+        move_form.partner_id = self.partner_a
+        move_form.invoice_date = invoice_1.invoice_date
+        move_form.ref = invoice_1.ref
+        with move_form.invoice_line_ids.new() as line_form:
+            line_form.product_id = self.product_a
+        with move_form.invoice_line_ids.new() as line_form:
+            line_form.product_id = self.product_b
+        invoice_2 = move_form.save()
+        self.assertRecordValues(invoice_2, [{'duplicated_ref_ids': invoice_1.ids}])
+
+    def test_in_invoice_multiple_duplicate_supplier_reference_batch(self):
+        """ Ensure duplicated ref are computed correctly even when updated in batch"""
+        invoice_1 = self.invoice
+        invoice_1.ref = 'a unique supplier reference that will be copied'
+        invoice_2 = invoice_1.copy(default={'invoice_date': invoice_1.invoice_date})
+        invoice_3 = invoice_1.copy(default={'invoice_date': invoice_1.invoice_date})
+
+        # reassign to trigger the compute method
+        invoices = invoice_1 + invoice_2 + invoice_3
+        invoices.ref = invoice_1.ref
+        self.assertRecordValues(invoices, [
+            {'duplicated_ref_ids': (invoice_2 + invoice_3).ids},
+            {'duplicated_ref_ids': (invoice_1 + invoice_3).ids},
+            {'duplicated_ref_ids': (invoice_1 + invoice_2).ids},
+        ])
+
+    def test_in_invoice_multiple_duplicate_supplier_reference_constrains(self):
+        """ Ensure that an error is raised on post if some invoices with duplicated ref share the same invoice_date """
+        invoice_1 = self.invoice
+        invoice_1.ref = 'a unique supplier reference that will be copied'
+        invoice_2 = invoice_1.copy(default={'invoice_date': invoice_1.invoice_date})
+        invoice_3 = invoice_1.copy(default={'invoice_date': invoice_1.invoice_date})
+
+        # reassign to trigger the compute method
+        invoices = invoice_1 + invoice_2 + invoice_3
+        invoices.ref = invoice_1.ref
+
+        # test constrains: batch without any previous posted invoice
+        with self.assertRaises(ValidationError):
+            (invoice_1 + invoice_2 + invoice_3).action_post()
+
+        # test constrains: batch with one previous posted invoice
+        invoice_1.action_post()
+        with self.assertRaises(ValidationError):
+            (invoice_2 + invoice_3).action_post()
+
+        # test constrains: single with one previous posted invoice
         with self.assertRaises(ValidationError):
-            invoice2.action_post()
+            invoice_2.action_post()
```

Comment:
The 2 first tests ensures it goes smoothly with the ORM style part of the `_fetch_duplicate_supplier_reference` method.
It ensures it works fine with and without the `NEWID`.

This test is testing the part that use the SQL query.

The different `assertRaise` test an error is raised if:
1. we try to post several invoices for which the duplicates concern themselves
2. we try to post several invoices for which a PREVIOUS duplicate already exists
    - [ ]  notes to myself: change the ref of invoice_3 otherwise it's also overlapping with (1)
3. we try to post one invoice for which a previous duplicate already exists

Reply:
There is no `NewId` involved in this test?
You are also talking about "previously posted" what does that mean? There is no difference?